### ISSUE TITLE
Potential issues with uploading to mapped drives in Windows

### ISSUE THREAD

**Issue by chetweewax** [njpanderson/push#20]
- Push version: 0.4.3
 - VS Code version: 1.23.1
 
I have set up a basic push.settings.json file as follows

{
	"service": "File",
	"File": {
		"root": "/Production"
	}
}

The file structure on my local server is essentially 

- D:\Server\ProductionBeta
- D:\Server\Production
- D:\Server\push.settings.json

I have two sites set up one with the root in each of those folders.
What am I missing, since I get the following error
"Cannot upload file "index.php" - It matches one of the defined ignoreGlobs filters."

**Comment by njpanderson** [comment]
The error you're getting usually occurs when a file matches the ignoreGlobs filters which are set at the workspace/settings level. Do you have the settings handy that you can paste in here? "index.php" is not normally ignored by default.

**Comment by chetweewax** [comment]
WorkSpace settings
{
	"folders": [
		{
			"path": "Production"
		},
		{
			"path": "ProductionTest"
		}
	],
	"settings": {
		"njpPush.ignoreGlobs": [
			"**/.DS_Store",
			"**/Thumbs.db",
			"**/desktop.ini",
			"**/.git/*",
			"**/.svn/*"
		],
		"files.exclude": {
			"**/.git": true,
			"**/.svn": true,
			"**/.hg": true,
			"**/CVS": true,
			"**/.DS_Store": true
		},
		"njpPush.queueCompleteMessageType": "status",
		"njpPush.settingsFilename": ".push.settings.json"
	}
}

**Comment by njpanderson** [comment]
Thanks, @chetweewax  - And just a few more points to clarify:

 - Is the settings file actually `push.settings.json` or was that just a mistype above? (it should start with a dot, according to your current settings).
 - What were you trying to upload when you received the above error?
 - Do you get this issue when uploading a single file (and does it matter if it's called "index.php")?
 - Are there any other `.push.settings.json` files?
 - I assume you're on Windows here - what version are you on?

Thanks!

**Comment by njpanderson** [comment]
Thanks @chetweewax - I'll have a look into it. I suspect this is a Windows specific issue. Will keep you posted!

**Comment by chetweewax** [comment]
Just as another note, when i right click a file and choose "compare with remote" I get the error "Directory is outside of root and cannot be created."

**Comment by njpanderson** [comment]
That might be due to the file service `root` property not being valid. In windows, would the root path for file not need be something more like `X:\SomeFolder\SomeOtherFolder`? Is the target root another windows drive?

**Comment by njpanderson** [comment]
Hey @chetweewax - I've been unable to replicate this using Windows 10 and the latest vscode/push installs.

Would you do me a favour and send me a screenshot of the layout of your project, including the push settings file (removing any sensitive data), as well as confirm the above — that the destination root path is actually accessible?

**Comment by AceTheWiz** [comment]
Hi. I have a similar challenge with the File service.
Win 7 x64, vscode v1.23.1
My push.settings are
```javascript
{
	"service": "File",
	"File": {
		"collisionUploadAction":"overwrite",
		"root": "Y:/bms"
	}
}
```
My local folder is
`C:\Projects\BMS\website`
Remote folder is a mapped network drive.
`y:\bms\`
The file i'm uploading
`C:\Projects\BMS\website\SGD\Processes\Production\Operations_South_Africa\Manufacturing\LRU_Production\LRU+Build+History\41587_LRU_Build_History_SGD.htm`
I get this error
`⚠️ EEXIST: file already exists, mkdir 'y:\bms\SGD'`
If I  "compare with remote" I also get the error 
`"Directory is outside of root and cannot be created."`

Hope this helps. Thanks.

**Comment by njpanderson** [comment]
Thanks guys - will look into this further when I get a chance tonight.

**Comment by njpanderson** [comment]
Have had another go on this using Windows 10:

A workspace path of the following:
```
C:\Users\IEUser\Documents\push-test\
├── file-transfer\
│   ├── .push.settings.json
│   └── test-file.txt
└── (other files)
```

The contents of `.push.settings.json` is:
```
{
    "service": "File",
    "File": {
        // "root": "C:\\Users\\IEUser\\Documents\\push-test-file-test",
        // "root": "E:\\Projects\\push-test-windows",
        "root": "E:/Projects/push-test-windows"
    }
}
```

All three variations of `root` seemed to upload `test-file.txt` to the location `E:\Projects\push-test-windows` successfully in my case. In this test, `E:\\` is a mapped drive.

Can anyone give a similar scenario a go and see if the error occurs with them as well?

**Comment by AceTheWiz** [comment]
I did some debugging and found this:
1. The copy works just fine if the destination folder is empty (i.e. No sub folders or files)
2. Subsequently when trying to "overwrite" destination folders and files, it fails with the "EEXIST" error. (This is actually expected BUT I have set the 'collisionAction' option to "overwrite" which is supposed to handle the EEXIST error as I understand it ?)

I also found a bug: There is a parameter mismatch in the transfer function.

When calling the transfer() from File.js line 63 there is an extra '>>' parameter, which ends up in the collisionAction parameter.

I made a fix thinking this would solve it but I still get the "EEXIST" error even with the  'collisionAction' option set to "overwrite".

Not an expert but I could not find where the  'collisionAction' option is actually used.

**Comment by njpanderson** [comment]
Thanks for the extra details, @AceTheWiz - I've had a look and made some minor changes to the way file caching, comparison and transfer works in and around the File service module.

If you could, would you be so kind as to try out the latest version within the `dev` branch and see if you can reproduce the issues please?

**Comment by AceTheWiz** [comment]
Great thanks, Will do.

**Comment by AceTheWiz** [comment]
Looking good ! Just one issue still. The extra parameter is still there.
This causes the collisionUploadAction option to be ignored so essentially it opens the user prompt regardless.

https://github.com/njpanderson/push/blob/b27976fe1e6532625a91228e3ec0efdfe54e2ff5/src/services/File.js#L50

When I remove this line it works fine.

Looking forward to update. Thanks for your time and effort.

**Comment by njpanderson** [comment]
Good spot - that's removed from there too. Fancy having another go from the `dev` branch?

**Comment by njpanderson** [comment]
This is hopefully fixed in release v0.4.4. If anyone else has further problems, please feel free to create a new issue. Thanks all!
